src/gs-update-monitor.c

```diff
@@ -500,6 +459,72 @@ schedule_update (GsUpdateMonitor *monitor,
 				       schedule_entry_scheduled_cb, helper);
 }
 
+static void
+schedule_updates_real (GsUpdateMonitor *monitor,
+		       GsAppList *apps_to_update)
+{
+	/* we have to use an updates counter to be able to know when all
+	 * updates have been scheduled */
+	monitor->num_scheduled_updates = gs_app_list_length (apps_to_update);
+	for (guint i = 0; i < monitor->num_scheduled_updates; i++) {
+		GsApp *app = gs_app_list_index (apps_to_update, i);
+		schedule_update (monitor, app,
+				 monitor->scheduled_updates_cancellable);
+	}
+}
+
+static void
+monitor_clear_scheduler (GsUpdateMonitor *monitor)
+{
+	g_clear_object (&monitor->scheduler);
+}
+
+static void
+scheduler_ready_cb (GObject *source_object,
+		    GAsyncResult *result,
+		    gpointer data)
+{
+	MwscScheduler *scheduler;
+	g_autoptr(GError) local_error = NULL;
+	g_autoptr(SchedulerHelper) helper = (SchedulerHelper *) data;
+	GsUpdateMonitor *monitor = helper->monitor;
+
+	scheduler = mwsc_scheduler_new_finish (result, &local_error);
+
+	if (scheduler == NULL) {
+		g_warning ("Error getting Mogwai Scheduler: %s", local_error->message);
+		return;
+	}
+
+	g_signal_connect_object (scheduler, "invalidated",
+				 (GCallback) monitor_clear_scheduler,
+				 monitor,
+				 G_CONNECT_SWAPPED);
+	monitor->scheduler = scheduler;
+
+	schedule_updates_real (monitor, helper->apps_to_update);
+}
+
+static void
+schedule_updates (GsUpdateMonitor *monitor,
+		  GsAppList *apps_to_update)
+{
+	/* if we don't have a valid scheduler yet, create it asynchronously and
+	 * delegate the updates scheduling to its callback */
+	if (monitor->scheduler == NULL) {
+		SchedulerHelper *helper = g_new0 (SchedulerHelper, 1);
+		helper->monitor = monitor;
+		helper->apps_to_update = apps_to_update;
+
+		mwsc_scheduler_new_async (monitor->scheduled_updates_cancellable,
+					  (GAsyncReadyCallback) scheduler_ready_cb,
+					  helper);
+		return;
+	}
+
+	schedule_updates_real (monitor, apps_to_update);
```

Comment:
I think `apps_to_update` is leaked on this path.

I think it probably makes more sense to not transfer ownership of it in to this function, and remove the `g_steal_pointer()` call from line 591 below.

Reply:
I need to transfer ownership because this will be used by the scheduler async cb. Thanks for spotting the leak.